### PR TITLE
[JBPM-9541] Kafka extension to support custom format

### PR DESCRIPTION
**Pull request by fjtirado**
Refactoring into several classes
Adding possibility of supporting more message formats that are not cloud events. 

**JIRA**: 

[JBPM-9541](https://issues.redhat.com/browse/JBPM-9541)

### CODE REVIEW

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+import org.jbpm.services.api.model.ProcessDefinition;
+import org.jbpm.services.api.model.SignalDesc;
+import org.jbpm.services.api.model.SignalDescBase;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.kie.server.services.jbpm.kafka.KafkaServerExtension.KAFKA_EXTENSION_PREFIX;
+import static org.kie.server.services.jbpm.kafka.KafkaServerExtension.processMessages;
+import static org.kie.server.services.jbpm.kafka.KafkaServerExtension.processSignals;
+import static org.kie.server.services.jbpm.kafka.KafkaServerExtension.topicFromSignal;
```
Comment (elguardian) [inline kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java:57]: move this to an utility class

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+                                                      Collection<T> signals) {
+        for (T signal : signals) {
+            if (!signal.getIncomingNodes().isEmpty()) {
+                topic2SignalBase.computeIfAbsent(topicFromSignal(signal), k -> new HashMap<>()).computeIfAbsent(
+                        signal, k -> new ArrayList<>()).add(deploymentId);
+            }
+        }
+    }
+
+    private <T extends SignalDescBase> void removeTopics(Map<String, Map<T, Collection<String>>> topic2SignalBase,
```
Comment (elguardian) [inline kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java:185]: move this to a topic registry class

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+    private void updateTopics(String deploymentId, ProcessDefinition processDefinition) {
+        if (processSignals()) {
+            addTopics(topic2Signal, deploymentId, processDefinition.getSignalsDesc());
+        }
+        if (processMessages()) {
+            addTopics(topic2Message, deploymentId, processDefinition.getMessagesDesc());
+        }
+    }
+
+    private void removeTopics(String deploymentId, ProcessDefinition processDefinition) {
```
Comment (elguardian) [inline kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java:169]: move this to a topic registry class

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+                consumer.wakeup();
+                subscribe(topic2Register);
+                isSubscribedCond.signal();
+            }
+        } finally {
+            changeRegistrationLock.unlock();
+        }
+    }
+
+    private void updateTopics(String deploymentId, ProcessDefinition processDefinition) {
```
Comment (elguardian) [inline kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java:160]: move this to a topic registry class

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+                new Thread(this).start();
+            } else {
+                consumer.wakeup();
+                subscribe(topic2Register);
+                isSubscribedCond.signal();
+            }
+        } finally {
+            changeRegistrationLock.unlock();
+        }
+    }
```
Comment (elguardian) [inline kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java:158]: move this to a topic registry class

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java (hunk 1 above)
Comment (fjtirado) [reply kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java]: done

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java (hunk 3 above)
Comment (fjtirado) [reply kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java]: done

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java (hunk 4 above)
Comment (fjtirado) [reply kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java]: done

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java (hunk 5 above)
Comment (fjtirado) [reply kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java]: Partially done. 
I think locks should be handled by consumer class. 
We can revisit this later.

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java (hunk 2 above)
Comment (fjtirado) [reply kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java]: done

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java (hunk 5 above)
Comment (fjtirado) [reply kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerConsumer.java]: Done

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java [hunk 6]
```diff
@@ -347,6 +356,17 @@ public String marshall(Object objectInput) {
         }
     }
 
+    @Override
+    public byte[] marshallAsBytes(Object objectInput) {
```
Comment (elguardian) [inline kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java:360]: why is this required ?

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java [hunk 7]
```diff
@@ -360,6 +380,19 @@ public String marshall(Object objectInput) {
         }
     }
 
+    @Override
+    public <T> T unmarshall(byte[] serializedInput, Class<T> type) {
```
Comment (elguardian) [inline kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java:384]: why is this required ?

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/DefaultEventProcessorFactory.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    private Map<ClassLoader, KafkaEventReader> readerInstanceMap = new ConcurrentHashMap<>();
+
+    private Map<String, KafkaEventWriter> topicWriterMap = new ConcurrentHashMap<>();
+    private Map<String, KafkaEventWriter> writerInstanceMap = new ConcurrentHashMap<>();
+
+    private AtomicBoolean defaultWriterInitialized = new AtomicBoolean(false);
+    private KafkaEventWriter defaultWriterInstance;
+
+    public KafkaEventReader getEventReader(String topic, ClassLoader cl) {
+        return topicReaderMap.computeIfAbsent(topic, t -> buildReader(topic, cl));
```
Comment (elguardian) [inline kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/DefaultEventProcessorFactory.java:39]: you need here to keep in mind that topic is singleton and it could have several class loaders ( different kjars)

**Review by elguardian — CHANGES_REQUESTED** [review]
some changes required regarding class loading that could cause strange bugs

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java (hunk 6 above)
Comment (fjtirado) [reply kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java:360]: this is used in RawJsonWriter. It is not strictly needed, but save a byte[]-string conversion, which provide some peformance gain

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java (hunk 7 above)
Comment (fjtirado) [reply kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java:384]: this is used in RawJsonReader It is not strictly needed, but save a byte[]-string conversion, which provide some peformance gain

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/DefaultEventProcessorFactory.java (hunk 8 above)
Comment (fjtirado) [reply kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/DefaultEventProcessorFactory.java]: good catch, thats a bug, I fixed by using a nested map (purpose is to avoid creating potential heavyweight classloader objects)

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java (hunk 6 above)
Comment (elguardian) [reply kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java:360]: move to the marshaller interface with default implementation UnsupportedOperationException

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java (hunk 7 above)
Comment (elguardian) [reply kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java:384]: move to the marshaller interface with default implementation UnsupportedOperationException

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-sample/src/test/java/org/kie/server/springboot/samples/KafkaFixture.java [hunk 9]
```diff
@@ -230,7 +228,7 @@ protected Properties producerProps(String bootstrapServer, boolean transactional
     }
     
     protected ListAppender<ILoggingEvent> addLogAppender() {
-        Logger logger = (Logger) LoggerFactory.getLogger(KafkaServerExtension.class);
+        Logger logger = (Logger) LoggerFactory.getLogger(KafkaServerExtension.class.getPackage().getName());
```
Comment (elguardian) [inline kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-sample/src/test/java/org/kie/server/springboot/samples/KafkaFixture.java:231]: this change is not needed and goes against all the other loggers. remove it.

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/RawJsonEventWriter.java [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+import org.kie.api.runtime.process.ProcessInstance;
+
+import static org.kie.server.services.jbpm.kafka.KafkaJsonUtils.jsonMarshallerWriter;
+
+
+public class RawJsonEventWriter implements KafkaEventWriter {
+
+    @Override
+    public byte[] writeEvent(ProcessInstance processInstance, Object value) throws IOException {
+        return jsonMarshallerWriter.marshallAsBytes(value);
```
Comment (elguardian) [inline kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/RawJsonEventWriter.java:28]: during the building of the write you can instantiate the json marshaller if you like with the right class loader. Don't use of statics.

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-sample/src/test/java/org/kie/server/springboot/samples/KafkaFixture.java (hunk 9 above)
Comment (fjtirado) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-sample/src/test/java/org/kie/server/springboot/samples/KafkaFixture.java:231]: I have to change that because we are using per logger class and test was expecting logs in KieServer

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/RawJsonEventWriter.java (hunk 10 above)
Comment (fjtirado) [reply kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/RawJsonEventWriter.java:28]: RawJsonEventWriter will use ClassLoader associated to the thread in which is executed, This is intentional and will work, I do not get the issue

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java (hunk 6 above)
Comment (fjtirado) [reply kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java:360]: it is in the marshaller with default implementation to call the string one

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java (hunk 7 above)
Comment (fjtirado) [reply kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java:384]: it is int he marshaller with default implementation to call the string one

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-sample/src/test/java/org/kie/server/springboot/samples/KafkaFixture.java (hunk 9 above)
Comment (gmunozfe) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-sample/src/test/java/org/kie/server/springboot/samples/KafkaFixture.java:231]: @elguardian In this case, the purpose of the test is to check that the exception is logged, no matter which the logger is associated with, so I think the change is valid after refactoring the extension

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/RawJsonEventWriter.java (hunk 10 above)
Comment (fjtirado) [reply kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/RawJsonEventWriter.java]: I changes the static marshaller  instance shared by RawJsonWriter and CloudEventWriter to private fields.

**Comment by fjtirado** [comment]
jenkins retest this

**Review by afalhambra — CHANGES_REQUESTED** [review]
Just a few comments

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/CloudEventReader.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+import org.kie.server.api.marshalling.MarshallingFormat;
+
+import static org.kie.server.services.jbpm.kafka.KafkaJsonUtils.mapper;
+
+class CloudEventReader implements KafkaEventReader {
+    private ClassLoader cl;
+
+    private Map<MarshallingFormat, Marshaller> marshallers = new EnumMap<>(MarshallingFormat.class);
+
+    public CloudEventReader(ClassLoader cl) {
```
[the remark this answers is not in the document]
Comment (fjtirado) [reply kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/CloudEventReader.java:33]: Thats fine, I want to stress that the constructor in public inside the package

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java [hunk 12]
```diff
@@ -174,7 +174,17 @@ public void setWriteNull(boolean writeNull) {
         EXTENSIONS = Collections.unmodifiableList(loadedPlugins);
     }
 
+    public JSONMarshaller(boolean formatDate) {
+        this(null, null, formatDate);
+    }
+
     public JSONMarshaller(Set<Class<?>> classes, ClassLoader classLoader) {
+        this(classes, classLoader, Boolean.parseBoolean(System.getProperty("org.kie.server.json.format.date",
+                "false")));
```
[the remark this answers is not in the document]
Comment (fjtirado) [reply kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java:183]: I did not want to change unrelated code, but lets do it

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerProducer.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(KafkaServerProducer.class);
+
+    // Kafka producer
+    private Producer<String, byte[]> producer;
+    private Supplier<Producer<String, byte[]>> producerSupplier;
+    private KafkaEventProcessorFactory factory;
+
+    private AtomicBoolean producerReady = new AtomicBoolean();
+
+    public KafkaServerProducer(KafkaEventProcessorFactory factory,
```
[the remark this answers is not in the document]
Comment (fjtirado) [reply kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerProducer.java:48]: Same comment than before, public is fine there

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaJsonUtils.java [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+
+import static org.kie.server.services.jbpm.kafka.KafkaServerUtils.KAFKA_EXTENSION_PREFIX;
+
+class KafkaJsonUtils {
+    
+    private KafkaJsonUtils() {}
+
+    public static final ObjectMapper mapper = new ObjectMapper()
```
[the remark this answers is not in the document]
Comment (fjtirado) [reply kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaJsonUtils.java:29]: same, I want things tha are used outside the package class to be public

File: kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerRegistration.java [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+import org.jbpm.services.api.model.SignalDesc;
+import org.jbpm.services.api.model.SignalDescBase;
+
+import static org.kie.server.services.jbpm.kafka.KafkaServerUtils.processMessages;
+import static org.kie.server.services.jbpm.kafka.KafkaServerUtils.processSignals;
+import static org.kie.server.services.jbpm.kafka.KafkaServerUtils.topicFromSignal;
+
+class KafkaServerRegistration {
+    private Map<String, Map<SignalDesc, Collection<String>>> topic2Signal = new HashMap<>();
+    private Map<String, Map<MessageDesc, Collection<String>>> topic2Message = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (fjtirado) [reply kie-server-parent/kie-server-services/kie-server-services-kafka/src/main/java/org/kie/server/services/jbpm/kafka/KafkaServerRegistration.java:39]: both maps needs to be updated at the same time.

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java [hunk 16]
```diff
@@ -360,6 +380,19 @@ public String marshall(Object objectInput) {
         }
     }
 
+    @Override
+    public <T> T unmarshall(byte[] serializedInput, Class<T> type) {
+
+        try {
+            Class<?> actualType = classesSet.contains(type) ? Object.class : type;
```
[the remark this answers is not in the document]
Comment (fjtirado) [reply kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/json/JSONMarshaller.java:387]: hmmmm, not sure about this one, this was copy paste of existing unmarshall, changing that might have unexpected effects. 
I will open a different JIRA, since this seems a bug to me

**Review by afalhambra — APPROVED** [review]
Amazing job - well done @fjtirado

**Comment by fjtirado** [comment]
jenkins retest this

**Comment by fjtirado** [comment]
@elguardian its greew now, I think we are ready to go

**Comment by fjtirado** [comment]
jenkins test this

**Comment by fjtirado** [comment]
unrelated failures, but jenkins test this

**Comment by fjtirado** [comment]
@elguardian green and ready to go

File: kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-sample/src/test/java/org/kie/server/springboot/samples/KafkaFixture.java (hunk 9 above)
Comment (elguardian) [reply kie-spring-boot/kie-spring-boot-samples/kie-server-spring-boot-sample/src/test/java/org/kie/server/springboot/samples/KafkaFixture.java:247]: fine. it does belong to tests not really prod code

**Review by elguardian — CHANGES_REQUESTED** [review]
Hi I still don't see how you are dealing with the problem that event reader depends on class loader. Seems to me it is not fixed yet
.Also keep in mind the clean up in those cases.
topic 1:M classloaders (deployments)

**Comment by fjtirado** [comment]
@elguardian ready to go... I hope

**Review by elguardian — APPROVED** [review]
LGTM
